### ISSUE TITLE
[kubelet] ReadinessProbe broken (failures are triggering kubelet to kill containers)

### ISSUE THREAD

**Issue by njuicsgz** [GoogleCloudPlatform/kubernetes#9649]
1. I have a RC with one pod whose container has readinessProbe.sh
2. The service run in container can not be 'ready' until about 2 min, so readinessProbe.sh exit with 1.
3. kubelet execute readinessProbe.sh every 10s by default, so it always get a status of 'False'
4. kubelet will kill and restart the container
5. goto 2, so it run into a loop of failure 

I remembered that readinessProbe fail does not cause container restart in v0.15.0, however livenessProbe will. And I checked the design in pod-states.md which keeps same behavior as before. So I am sure it is a bug in v0.18.1.

```
root@pmd-122:~# kubectl version
Client Version: version.Info{Major:"0", Minor:"18", GitVersion:"v0.18.1", GitCommit:"befd1385e5af5f7516f75a27a2628272bb9e9f36", GitTreeState:"clean"}
Server Version: version.Info{Major:"0", Minor:"18", GitVersion:"v0.18.1", GitCommit:"befd1385e5af5f7516f75a27a2628272bb9e9f36", GitTreeState:"clean"}
```

Error Log
```
I0610 16:11:50.787896    2664 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"fb-pmd", Name:"walle-java-vpaas.test.001-2r93u", UID:"54d2f9ef-0f48-11e5-9420-0050568104ec", APIVersion:"v1beta3", ResourceVersion:"605618", FieldPath:"spec.containers{walle-java}"}): reason: 'pulled' Successfully pulled image "xa.repo.ndp.com:5000/facebook_pmd/walle-java:vpaas.test.001"
I0610 16:11:50.836952    2664 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"fb-pmd", Name:"walle-java-vpaas.test.001-2r93u", UID:"54d2f9ef-0f48-11e5-9420-0050568104ec", APIVersion:"v1beta3", ResourceVersion:"605618", FieldPath:"spec.containers{walle-java}"}): reason: 'created' Created with docker id ea151bc580280b13da95d88239641cf9d8a9cdb795b956c6981b97bc5d47b110
I0610 16:11:50.917270    2664 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"fb-pmd", Name:"walle-java-vpaas.test.001-2r93u", UID:"54d2f9ef-0f48-11e5-9420-0050568104ec", APIVersion:"v1beta3", ResourceVersion:"605618", FieldPath:"spec.containers{walle-java}"}): reason: 'started' Started with docker id ea151bc580280b13da95d88239641cf9d8a9cdb795b956c6981b97bc5d47b110
I0610 16:11:54.905371    2664 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"fb-pmd", Name:"walle-java-vpaas.test.001-2r93u", UID:"54d2f9ef-0f48-11e5-9420-0050568104ec", APIVersion:"v1beta3", ResourceVersion:"605618", FieldPath:"spec.containers{walle-java}"}): reason: 'unhealthy' Readiness probe failed: not ok
I0610 16:11:54.905050    2664 manager.go:1407] pod "walle-java-vpaas.test.001-2r93u_fb-pmd" container "walle-java" is unhealthy (probe result: failure), it will be killed and re-created.
```

**Comment by mikedanese** [comment]
Oh uh. This is a definitely a bug. Thanks for submitting the issue!

**Comment by mikedanese** [comment]
Second time we've had this regression. :(

Break: #6327
Fix: #6327
Break: #8238

Docs on intended behavior [here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/pod-states.md#container-probes)

**Comment by mikedanese** [comment]
It seems the [pkg/kubelet/prober/prober.go.prober.Probe()](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L94) method has been a bit confusing for maintainers and the monolithic method primarily exists because of the semantic that links readinessProbe to livenessProbe: "if liveness probe is unsuccessful, then readiness is false". Otherwise, the readinessProbe and livenessProbe could be split up.

On top of the extra complexity in the code, I don't think this semantic is correct from a UX standpoint. Take for example a pod with two containers: a webserver container with a liveness and readiness probe, and a log shipping container with a liveness probe. If the logging container livenessProbe fails, then the logging container's readiness will be set to false, then the pod readiness will be set to false, then the webserver will be removed from service endpoints even if the webserver container is live and ready. This seems incorrect.

Proposal: for v1, I'd like to the results of readinessProbe and livenessProbe completely independent (at least from an api standpoint).

If that's to big of a change for v1, I'll just fix the bug.

cc @bgrant0607 @thockin @dchen1107

**Comment by njuicsgz** [comment]
@mikedanese I am completely agree with you that we should separate liveness and readiness probe. However, I am also afraid that will work out too many code change.

I have a proposal and run in my env well now:
1. computePodContainerChanges define the keep/restart container list, container will be kept if probe.Success, or killing/restart in else. Since we have 3 status of probe, we should only kill/restart container if probe.Failure while just keep container if probe.Unknown.
2. We have "initialDelaySeconds" for probe handler. ReadinessProbe returns probe.Failure if the container running time less then "initialDelaySeconds", which will cause container killing/restart. I think it is too rough, I think probe.Unknown is a reasonable status since kubelet does not probe the container. And based on 1, container will not be kill/restarted in the period of "initialDelaySeconds" with status probe.Unknown.
3. It is hard for k8s to know how log the "initialDelaySeconds" should be, if the readiness probe still fail after that period the problem come again. So, we should provide a mechanism to let the readinessProbe aware the status of container service. We could define 3 types exit code which fully mapped 3 probe result, e.g. exit 0 means probe.Sucess; exit 1 means probe.Unknown; other exit codes means probe.Failure. So, the writer of script readinessProbe.sh could design how to process accordingly. Thus,
exit 0: Ready: True
exit 1: Ready: False, not restart container
exit 2+: Ready: False, restart container

Please review this proposal, I will add a PR if we achieve agreement on this.

**Comment by davidopp** [comment]
> If the logging container livenessProbe fails, then the logging container's readiness will be set to false,
> then the pod readiness will be set to false, then the webserver will be removed from service endpoints > even if the webserver container is live and ready. This seems incorrect.

I'm not sure this is incorrect. Putting containers into a pod means they work closely together, so I think you do want the behavior you described. For example, if the log saver stops working, you probably want to stop directing traffic to the web server it is paired with (since you are no longer able to save the logs from the web server). Likewise in the case of a "data puller" and web server, if the data puller stops working, you probably want to stop directing traffic to the web server since the web server will be serving stale data.

It does seem like a bug if readiness probe failure without liveness probe failure is causing container or pod restart, though.

**Comment by mikedanese** [comment]
@davidopp your use case is still possible (put a readiness probe on the log shipper or data puller), however if I want my webserver to stay up, even if my log shipper is failing a liveness probe, there's nothing I can do about it with the current semantics.

**Comment by mikedanese** [comment]
@njuicsgz We have a code freeze for 1.0 at the end of the week (tomorrow), and I think what you are proposing is a larger api change then just unwinding readiness from liveness. readinessProbe should kill a container but only if it's and execAction and it returns a status code > 1? That seems intuitive. I also think that the concept of readiness and liveness should be orthogonal in the api, and have a disjoint sets of actions that each trigger.

@davidopp I've discussed this with @dchen1107 and have a PR in flight that decouples readiness from liveness. Speak up or forever hold your peace (until v1.0-post)!

**Comment by davidopp** [comment]
I think the coupling is slightly better than the decoupling (new way requires people to write an extra readiness probe) but I'm fine with the change.

**Comment by njuicsgz** [comment]
So that readiness will only mark status and send event, but not report its result to SyncPod().
LGTM thanks.

**Comment by dchen1107** [comment]
I just merged #9680, close the issue. @njuicsgz Thanks for filing the issue and review the pr.
